hash_practice/exercises.py

```diff
@@ -2,18 +2,65 @@
 def grouped_anagrams(strings):
     """ This method will return an array of arrays.
         Each subarray will have strings which are anagrams of each other
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(n)
+        Space Complexity: O(n)
     """
-    pass
+    if len(strings) == 0:
+        return []
+
+    hash = {}
+    for string in strings:
+        #sort out the word alphabetically since we are working with anagrams and make that the key
+        #sorted returns a list and we need to change it back to a string by using join
+        key = ''.join(sorted(string))
+        if key not in hash:
+            #bracket notation to add it in as a list
+            #we're adding in the actual, not sorted, version of the string
+            hash[key] = [string]
+        else:
+            #since the list is already there, append to it
+            hash[key].append(string)
+    print(hash.values())
+    return hash.values()
 
 def top_k_frequent_elements(nums, k):
     """ This method will return the k most common elements
         In the case of a tie it will select the first occuring element.
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(n)
+        Space Complexity: O(n)
     """
-    pass
+    #edge case
+    if len(nums) == 0:
+        return []
+
+    #create a hashmap
+    hash = {}

```

Comment:
Minor style thing: while it's not a big deal for these small problems, usually it's worth trying to find more descriptive names for your variables than `hash`. A name like `hash` describes what the variable _is_, but it's usually more informative to pick a name like `num_freqs` that describes what the variable is _for_.

This happens a few other places in the code, but I'll only mention it once.